package uniflex_framework;

message Attribute {
    required string name = 1;
}
message Function {
    required string name = 1;
}
message Event {
    required string name = 1;
}
message Service {
    required string name = 1;
}


message Device {
    required uint32 id = 1;
    required string name = 2;
}

message Module {
    enum ModuleType {
      MODULE = 0;
      DEVICE = 1;
      APPLICATION = 2;
    }
    required string uuid = 1;
    required uint32 id = 2;
    required string name = 3;
    required ModuleType type = 4;
    optional Device device = 5;
    repeated Attribute attributes = 6;
    repeated Function functions = 7;
    repeated Event in_events = 8;
    repeated Event out_events = 9;
    repeated Service services = 10;
}

message NodeInfoMsg {
    required string agent_uuid = 1;
    required string ip = 2;
    required string name = 3;
    required string hostname = 4;
    optional string info = 5;
    repeated Module devices = 6;
    repeated Module modules = 7;
    repeated Module applications = 8;
}

message NodeInfoRequest {
    required string agent_uuid = 1;
}

message NodeAddNotification {
    required string agent_uuid = 1;
}

message NodeExitMsg {
    required string agent_uuid = 1;
    optional string reason = 2;
}

message HelloMsg {
    required string uuid = 1;
    required uint32 timeout = 2;
}
